### PR TITLE
#WIP: Google_Cloud_docker building block

### PR DESCRIPTION
**Pull request by DiSanchz**
WORK IN PROGRESS. Content has been reviewed, small typos have been corrected and all comments have been addressed by 14/10. 

@narendramukherjee
@shrabasteebanerjee 
@juliehabets 
@casruger

### CODE REVIEW

File: content/building-blocks/automate-and-execute-your-work/reproducible-work/google_cloud_docker.md
Comment (narendramukherjee) [inline content/building-blocks/automate-and-execute-your-work/reproducible-work/google_cloud_docker.md:89]: `In the same way that you needed a dockerfile to provide Docker with the necessary instructions to build an image, a docker-compose file tells the docker engine how to build container(s) from the image.`

Also, maybe we can add a note to say that `docker run` command can theoretically be used to create and run containers and do what docker-compose does. But the advantage offered by docker-compose is that it can very smoothly create multiple containers, allow them to communicate with each other or start up containers with complicated settings. Also having all those instructions in a file (instead of typing it on the terminal) makes it easier to reproduce.

File: content/building-blocks/automate-and-execute-your-work/reproducible-work/google_cloud_docker.md
Comment (narendramukherjee) [inline content/building-blocks/automate-and-execute-your-work/reproducible-work/google_cloud_docker.md:102]: Theoretically, this is `docker compose up <service-to-be-run>` or `docker compose up` if you want to run all the services in the docker-compose file

File: content/building-blocks/automate-and-execute-your-work/reproducible-work/google_cloud_docker.md
Comment (narendramukherjee) [inline content/building-blocks/automate-and-execute-your-work/reproducible-work/google_cloud_docker.md:128]: This is kinda important if you want to access the notebook again and again without rebuilding the container. That token usually is not shown again if you restart an old container. So the trick is to copy that token and *make a new password* at the bottom of the Jupyter log-in page. Then if you re-start the notebook, you just log in with the password.

This becomes even more important when you move to remote systems (eg. Google cloud) and have to re-access notebooks by restarting containers.

File: content/building-blocks/automate-and-execute-your-work/reproducible-work/google_cloud_docker.md
Comment (narendramukherjee) [inline content/building-blocks/automate-and-execute-your-work/reproducible-work/google_cloud_docker.md:2]: I feel like we should make 2 separate tutorials: one for Windows (and say it is explicitly for Windows) and one for the cloud (and by extension for Unix systems). Were you thinking of adding Google cloud to this same page @DiSanchz ? I think it'll be too long as a tutorial then. Will be great to hear from @shrabasteebanerjee on this issue too.

**Review by narendramukherjee — CHANGES_REQUESTED** [review]
@DiSanchz I have given a few comments and suggestions - look through them and incorporate changes as necessary. Feel free to give your thoughts in response too - you can reply to each review comment in the box below the comment.

File: content/building-blocks/automate-and-execute-your-work/reproducible-work/google_cloud_docker.md
Comment (DiSanchz) [reply content/building-blocks/automate-and-execute-your-work/reproducible-work/google_cloud_docker.md:16]: The original intention with this warning was to provide some context and express a few general ideas about the building block, posible future lines of expansion etc. It was not intended to be on display in the final version for TSH's readers but just for reviewers until the building block is more mature/finished.

In any case if any piece of information from here is finally deemed useful it can be keeped an integrated within the rest of the building block!

File: content/building-blocks/automate-and-execute-your-work/reproducible-work/google_cloud_docker.md
Comment (DiSanchz) [reply content/building-blocks/automate-and-execute-your-work/reproducible-work/google_cloud_docker.md:35]: Not really descriptive of the building block taking into account the content it will cover with all the next additions, this definitively has to be upgraded

File: content/building-blocks/automate-and-execute-your-work/reproducible-work/google_cloud_docker.md
Comment (DiSanchz) [reply content/building-blocks/automate-and-execute-your-work/reproducible-work/google_cloud_docker.md:74]: Got it!

File: content/building-blocks/automate-and-execute-your-work/reproducible-work/google_cloud_docker.md
Comment (DiSanchz) [reply content/building-blocks/automate-and-execute-your-work/reproducible-work/google_cloud_docker.md:102]: I will include the puntualization

File: content/building-blocks/automate-and-execute-your-work/reproducible-work/google_cloud_docker.md
Comment (DiSanchz) [reply content/building-blocks/automate-and-execute-your-work/reproducible-work/google_cloud_docker.md:128]: Perfect! The fragment will be expanded in that direction

File: content/building-blocks/automate-and-execute-your-work/reproducible-work/google_cloud_docker.md
Comment (DiSanchz) [reply content/building-blocks/automate-and-execute-your-work/reproducible-work/google_cloud_docker.md:2]: My intention for this week was indeed to make the modifications derived from the feedback and then add the google cloud segment on top.  In any case I also had this impression that its extension might be above average (once the google cloud section is added) when compared with other building blocks already available in TSH.

File: content/building-blocks/automate-and-execute-your-work/reproducible-work/google_cloud_docker.md
Comment (casruger) [inline content/building-blocks/automate-and-execute-your-work/reproducible-work/google_cloud_docker.md:3]: The ideal description lengt is around 145 characters. Yours is 113, so a bit more information would be nice. And in general it is better to describe what the reader would find on the page. In this case something like: "Use Docker to publish reproducible research. Learn the benefits of the docker container, how to set it up, and how you can …" This helps google read the page and helps us get even more visitors.

File: content/building-blocks/automate-and-execute-your-work/reproducible-work/google_cloud_docker.md
Comment (casruger) [inline content/building-blocks/automate-and-execute-your-work/reproducible-work/google_cloud_docker.md:8]: The first alias is used to make a short link. In this case I would make it something like: 

Aliases:

> - /run/jupyter-on-cloud

File: content/building-blocks/automate-and-execute-your-work/reproducible-work/google_cloud_docker.md
Comment (casruger) [inline content/building-blocks/automate-and-execute-your-work/reproducible-work/google_cloud_docker.md:3]: The ideal lenght of the description is 145 characters. This description is 76. An example of a longer description would be: Learn to install and set up Docker in a Google Cloud instance, build environments, and connect docker with Google Cloud storage buckets…

File: content/building-blocks/automate-and-execute-your-work/reproducible-work/google_cloud_docker.md
Comment (casruger) [inline content/building-blocks/automate-and-execute-your-work/reproducible-work/google_cloud_docker.md:178]: Mayeb make this a sub title since now the extra is just as important as the normal steps. So i would change the ### to ####

**Comment by casruger** [comment]
@DiSanchz Good building block. Made good use of the layout and style. I made some suggestions to make it fit more of our regular style. If you have looked at them and changed what you think is appropriat I will post it online.

**Comment by hannesdatta** [comment]
Good review, @casruger. Do we have review guidelines documented somewhere? I mean, maybe like a general comments people can copy paste to check? Just thinking about scaling here.

File: content/building-blocks/automate-and-execute-your-work/reproducible-work/google_cloud_docker.md
Comment (narendramukherjee) [inline content/building-blocks/automate-and-execute-your-work/reproducible-work/google_cloud_docker.md:74]: We need to mention that on some Linux distributions you need to add your username to the docker group to be able to run docker commands without sudo. https://docs.docker.com/engine/install/linux-postinstall/

File: content/building-blocks/automate-and-execute-your-work/reproducible-work/google_cloud_docker.md
Comment (narendramukherjee) [inline content/building-blocks/automate-and-execute-your-work/reproducible-work/google_cloud_docker.md:94]: Spelling error: `will be placed`

File: content/building-blocks/automate-and-execute-your-work/reproducible-work/google_cloud_docker.md
Comment (narendramukherjee) [inline content/building-blocks/automate-and-execute-your-work/reproducible-work/google_cloud_docker.md:161]: This error is probably related to not adding your username to the docker grp post-install (check comment I made above). chmod 666 gives everyone permission to mess around with the docker engine, which might not be the best idea. Could you check if adding your user to the docker grp solves this problem actually?

File: content/building-blocks/automate-and-execute-your-work/reproducible-work/google_cloud_docker.md
Comment (narendramukherjee) [inline content/building-blocks/automate-and-execute-your-work/reproducible-work/google_cloud_docker.md:213]: Spelling error: `immediately`

File: content/building-blocks/automate-and-execute-your-work/reproducible-work/google_cloud_docker.md
Comment (narendramukherjee) [inline content/building-blocks/automate-and-execute-your-work/reproducible-work/google_cloud_docker.md:248]: Possible typo: `port number` instead of `four number`

**Review by narendramukherjee — COMMENTED** [review]
@DiSanchz To me, this looks very close to the finished product now - I have just added a few new small comments after which I am ready to approve this. Can you take a look through these new comments (and also the comments from the other reviewers) and I will resolve the old comments I had made.
Also @shrabasteebanerjee can take a look and provide her approval too.

File: content/building-blocks/automate-and-execute-your-work/reproducible-work/google_cloud_docker.md
Comment (DiSanchz) [reply content/building-blocks/automate-and-execute-your-work/reproducible-work/google_cloud_docker.md:8]: Got it!

File: content/building-blocks/automate-and-execute-your-work/reproducible-work/google_cloud_docker.md
Comment (DiSanchz) [reply content/building-blocks/automate-and-execute-your-work/reproducible-work/google_cloud_docker.md:3]: Perfect! I've modified it to the 150 character long: "Learn how to import a containerized python environment with docker, run it on a google cloud instance and connect it with Google cloud storage buckets"

File: content/building-blocks/automate-and-execute-your-work/reproducible-work/google_cloud_docker.md
Comment (DiSanchz) [reply content/building-blocks/automate-and-execute-your-work/reproducible-work/google_cloud_docker.md:178]: That makes sense, done!

**Comment by DiSanchz** [comment]
> @DiSanchz Good building block. Made good use of the layout and style. I made some suggestions to make it fit more of our regular style. If you have looked at them and changed what you think is appropriat I will post it online.

Thanks for the comments @casruger! I am going to work on the last changes and corrections and I'll let you know once all of these have been  addressed!

File: content/building-blocks/automate-and-execute-your-work/reproducible-work/google_cloud_docker.md
Comment (DiSanchz) [reply content/building-blocks/automate-and-execute-your-work/reproducible-work/google_cloud_docker.md:248]: I wanted to make it easier to identify the port number for users for which this concept may be new so I decided to stress that it is 4-digit long. In any case there is a typo present as there should be an extra plural on "number" so the idea is correctly expressed. All in all I've decided to reformulate that fourth step to: 

_"4. In the section "Protocols and ports" check "TCP" and then in the box below introduce the four-digit number that identifies the port where Jupyter is being executed in your instance."_

This will be reflected in the next update of the pull request.

File: content/building-blocks/automate-and-execute-your-work/reproducible-work/google_cloud_docker.md
Comment (DiSanchz) [reply content/building-blocks/automate-and-execute-your-work/reproducible-work/google_cloud_docker.md:161]: That was indeed the case, after adding my username to the group this issue was solve. Consequentiall I have deleted this warning from the building block.

File: content/building-blocks/automate-and-execute-your-work/reproducible-work/google_cloud_docker.md
Comment (DiSanchz) [reply content/building-blocks/automate-and-execute-your-work/reproducible-work/google_cloud_docker.md]: Corrected!

**Comment by hannesdatta** [comment]
Who is reviewing this? Is my action required?

**Comment by shrabasteebanerjee** [comment]
> Who is reviewing this? Is my action required?

Hi @hannesdatta , @narendramukherjee and I are looking into this - it's all set to be merged once Cas' formatting related comments have been addressed. Then it's ready! No other action required on your end.

**Comment by casruger** [comment]
@DiSanchz  Looks great! If you change the pull request from draft to final I will merge them

**Comment by casruger** [comment]
@DiSanchz I have published it

**Pull request merged**
